### PR TITLE
Update Go modules

### PR DESCRIPTION
**Pull request by SuperQ**
* Update Go modules for new protobuf version.
* Update Go modules in sub projects.
* Enable dependabot for sub projects.

### CODE REVIEW

**Comment by SuperQ** [comment]
This is going to require a bunch of test refactoring.

**Comment by beorn7** [comment]
Please ping me again when the tests are passing. :)

**Comment by beorn7** [comment]
General idea looks OK. But there are even more code locations where we have utilized the text formatted protobuf, which still fail.

I guess you don't even need to change from angle bracket to curly braces with the new comparison helper.

Also, relevant lint warnings and now a merge conflict (which, I'm afraid, has to do with even more tests).

**Comment by SuperQ** [comment]
Yea, I'm not sure about what to do to fix a few of these, which is why I did only a partial update so far.
* Tests for scrape output that test text proto.
* Inline text proto errors in the text scrape format.

Ideas:
* Write a `stabletextproto` package.
* Take the scrape output and marshal it back and compare.

**Review by bwplotka — COMMENTED** [review]
Thanks for this! Generally looking good.

Indeed we could write small function to convert new output to old, but we might as well refactor as you do now in one go (thanks for amazing work!). Please make sure to keep Examples in, they are important to have example type (for documentation reasons).

> Tests for scrape output that test text proto.
> Inline text proto errors in the text scrape format.

Can you point us to problematic parts? Would love to take closer look on what's best approach there 🤗

File: prometheus/example_metricvec_test.go [hunk 1]
```diff
@@ -106,7 +106,7 @@ func (v *InfoVec) MustCurryWith(labels prometheus.Labels) *InfoVec {
 	return vec
 }
 
-func ExampleMetricVec() {
+func TestExampleMetricVec(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (SuperQ) [reply prometheus/example_metricvec_test.go:109]: Ahh, didn't know that.

**Comment by SuperQ** [comment]
@bwplotka When we print some errors in the registry output, we embed part of the protobuf as output.

For example [this section of registry.go](https://github.com/prometheus/client_golang/blob/5e78d5f66b851fef874b783814b2e884df2798d0/prometheus/registry.go#L887-L914). 

We check for errors like:
```
An error has occurred while serving metrics:

collected metric "name" { label:<name:"constname" value:"\377" > label:<name:"labelname" value:"different_val" > counter:<value:42 > } has a label named "constname" whose value is not utf8: "\xff"
```

**Comment by bwplotka** [comment]
Ok, but that is again - expecting in unit test a different format, so we just need to switch to new text format in [expected output](https://github.com/prometheus/client_golang/blob/5e78d5f66b851fef874b783814b2e884df2798d0/prometheus/registry_test.go#LL255C2-L255C45)? 🤔

**Comment by SuperQ** [comment]
@bwplotka No, it's not just that. The new protobuf `prototext` package explicitly randomizes the whitespace in the output. Making every test slightly different.

https://github.com/protocolbuffers/protobuf-go/blob/v1.30.0/internal/encoding/text/encode.go#L226-L228

**Comment by SuperQ** [comment]
Let me try calling `detrand.Disable()` explicitly to see if that helps.

**Comment by beorn7** [comment]
For the Go example use case, the non-deterministic text output is really annoying. We cannot even implement our own comparison function (like one that ignores whitespace).

**Comment by bwplotka** [comment]
So there is some truth that tests [comparing text encoding is a little brittle and blocks serialization innovations (?)](https://github.com/golang/protobuf/issues/1121#:~:text=benefit%20the%20majority.-,Golden%20tests,-The%20most%20common). If we follow suggestions I think we have to get rid of checking exact errors and parse text format or proto format to structure and compare semantically those (e.g. [here](https://github.com/prometheus/client_golang/blob/5e78d5f66b851fef874b783814b2e884df2798d0/prometheus/registry_test.go#L569))

**Comment by SuperQ** [comment]
Closing in favor of #1323

**Pull request closed without merging**